Add a tides test case for the VR45to5 mesh

sbrus89 opened the pull request:
This PR adds the workflow to generate the mesh and initial conditions for the VR45to5 barotropic tides mesh, run the simulation, and perform the global tidal constituent error analysis. It relies on the developments from https://github.com/E3SM-Project/E3SM/pull/6310. This case reproduces the MPAS-Ocean configuration that gives the best deep water M2 RMSE accuracy (3.3 cm compared to TPXO9) to date.

<!--
Below are a few things we ask you or your reviewers to kindly check. 
***Remove checks that are not relevant by deleting the line(s) below.***
-->
Checklist
* [x] User's Guide has been updated
* [x] Developer's Guide has been updated
* [x] API documentation in the Developer's Guide (`api.rst`) has any new or modified class, method and/or functions listed
* [x] Documentation has been [built locally](https://mpas-dev.github.io/compass/latest/developers_guide/building_docs.html) and changes look as expected
* [x] The `E3SM-Project` submodule has been updated with relevant E3SM changes
* [x] Document (in a comment titled Testing in this PR) any testing that was used to verify the changes
<!--
Please note any issues this fixes using closing keywords: https://help.github.com/articles/closing-issues-using-keywords
-->

sbrus89 committed:
Add VR45to5 tides case

sbrus89 committed:
Updates to VR tides case

sbrus89 commented:
**Testing**

The mesh generated looks like:
![mesh_EC](https://github.com/MPAS-Dev/compass/assets/3514023/e86ad137-1eb5-4dd8-9c2f-34e7402bec04)

Land ice draft is included in the init step
![ice_draft](https://github.com/MPAS-Dev/compass/assets/3514023/073cb448-34f5-4c89-86c8-30de5afe79ca)

Bathymetric slopes and buoyancy frequency climatology are calculated/interpolated in the init step for topographic wave drag parameterizations:
![bathy_slope](https://github.com/MPAS-Dev/compass/assets/3514023/93f1feb9-0462-45d7-9d75-40d297ccfb32)
![Nb](https://github.com/MPAS-Dev/compass/assets/3514023/d337bd60-2a66-4046-86a4-ca9c67e885eb)

The harmonic constituents are compared against TPXO9 in the analysis step:
![M2_plot](https://github.com/MPAS-Dev/compass/assets/3514023/5be04e17-28b9-4e63-9e22-6cf7acc7b35a)
![N2_plot](https://github.com/MPAS-Dev/compass/assets/3514023/b765cd02-22c6-4e7e-b83e-19c62730b435)
![S2_plot](https://github.com/MPAS-Dev/compass/assets/3514023/df835f82-74eb-44a1-8815-9869bda0214a)
![O1_plot](https://github.com/MPAS-Dev/compass/assets/3514023/d78f9214-9b38-42d8-a8c0-9aac01f3a6f9)
![K1_plot](https://github.com/MPAS-Dev/compass/assets/3514023/f7146d28-c1b9-4b03-aca8-97bfbd796616)

sbrus89 commented on compass/ocean/tests/tides/mesh/vr45to5/__init__.py:
```diff
@@ -0,0 +1,284 @@
+# from skimage.filters.rank import median, percentile
+
+
+class VRTidesMesh(SphericalBaseStep):
```
This should probably inherit from `QuasiUniformSphericalMeshStep` then I can reuse the `setup` and `run` methods (if I refactor the return variables from `build_cell_width_lat_lon` and the input variables to `make_jigsaw_mesh`

xylar replied:
If possible, that does seem better.  But we would want to avoid updating `QuasiUniformSphericalMeshStep` if it can be avoided.

dengwirda commented on compass/ocean/tests/tides/dem/__init__.py:
```diff
@@ -0,0 +1,59 @@
+        super().__init__(test_case=test_case, name='pixel',
+                         ntasks=1, min_tasks=1, openmp_threads=1)
+
+        self.add_input_file(
```
Building these DEM datasets could be done once and cached somewhere, as it's pretty expensive and is typically only redone when GEBCO release a new bathymetry (once per year).

dengwirda commented on compass/ocean/tests/tides/dem/dem_pixel.py:
I'll keep these tools updated in https://github.com/dengwirda/dem for the latest GEBCO etc.

dengwirda commented on compass/ocean/tests/tides/forward/namelist.ocean:
```diff
@@ -4,10 +4,14 @@ config_calendar_type = 'noleap'
 config_time_integrator = 'RK4'
+config_use_mom_del2 = .true.
+config_mom_del2 = 3.333
+config_use_mom_del4 = .true.
```
I didn't do a lot of tuning on the del^k dissipation, so this could be looked at in the future if there's interest.

dengwirda commented on compass/ocean/tests/tides/init/initial_state.py:
Since the model is BTR-only, the following is a simple non-iterative adjustment of the SSH due to ISC pressure.

dengwirda reviewed (commented):
Thanks @sbrus89 this runs for me successfully on pm-cpu using https://github.com/E3SM-Project/E3SM/pull/6310. A few things I noticed:
- It looks like the plots show TPXO8 rather than TPXO9?
- It looks like the wave-drag step may be printing quite a bit of output to the logs which could be suppressed? Also some long python loops which may be expensive.
- You may see improved performance if jigsaw is upgraded to a more recent version that runs with multi-threading + accelerated steepest descent.

dengwirda commented:
Not relevant to this PR specifically, but since the errors in M2 are quite competitive here, I wonder if tuning for other constituents too (RMSE_total: https://gmd.copernicus.org/articles/14/1125/2021/) may make sense in the future, so as not to over fit on M2.

sbrus89 committed:
Fix analysis plot titles and flake8 issues

sbrus89 committed:
Remove log file output and add docstrings to init

mark-petersen reviewed (approved):
Approving based on @sbrus89's testing above. Thank you! It is great to have this published test case in the repository!

sbrus89 committed:
Use bed derivatives from pixel file in calculate_wave_drag.py

sbrus89 committed:
Inherit from QuasiUniformSphericalMeshStep for VRTidesMesh class

sbrus89 committed:
Update documentation

xylar commented:
@sbrus89, as I just mentioned on slack, my attempt to update the submodule was not very successful.  https://github.com/E3SM-Project/E3SM/pull/6310 seems to be causing segfaults on Chrysalis with Intel and OpenMPI.

sbrus89 commented:
@xylar, the segfault issues should be fixed via #825, which turns off topographic wave drag for the global ocean test cases. We should only test `config_use_topographic_wave_drag=.true.` for tides cases like the one in this PR.

xylar commented:
Almost there, just waiting on #827

xylar commented:
@sbrus89, now that #827 is in, can you rebase and test one last time?  Then, I'll merge.

sbrus89 commented:
@xylar, yes I will do that. Thanks for helping with the submodule update!

xylar commented:
Following up from a month ago, a reminder to rebase and retest.

sbrus89 committed:
Add remap topography step to tides mesh step

sbrus89 committed:
Improve tides analysis step

 - Run extract_HC in parallel

 - Add list of constituents to analyze to config file

sbrus89 commented:
@xylar, sorry for the big delay on this. Everything is working for me after the rebase now that I sorted out the issues with the ice shelf cavity culling step. I'm retesting the Icos7 test to make sure it is still working as expected. I'll let you know once I've verified that it is.

cblakely97 committed:
Fix minor bugs in tides test case

   - Adds remap topography step to tides case to update cull_mesh
     behavior

sbrus89 committed:
Separate extraction and plotting in tide analysis

bmooremaley committed:
Update tides config files for tempest and constituents.

bmooremaley committed:
Reverted manual used of extract_HC in ocean/tides/analysis
